### ISSUE TITLE
Pagination with subquery broken

### ISSUE THREAD

**Issue by bkilshaw** [laravel/scout#450]
- Scout Version: 8.6.0
- Laravel Version: 8.29.0
- PHP Version: 7.4.15
- Database Driver & Version: Meilesearch 0.19.0

### Description:

The paginator breaks when you append an additional query on the results.

### Steps To Reproduce:

Create 10 users all with the name "John". Give five of them with a location_id of 1 and the other five a location_id of 2.

```
$users = User::search('John')->paginate(5);
$users->count(); // 5
```

```
$users = User::search('John')->query(function($q) {
    $q->where('location_id', 2);
})->paginate(5);
$users->count(); // 5 - (matches where location_id != 2)
```

What appears to be happening is that the results are paginated based on the `search()` results and ignores the `query()`. Meanwhile, the actual pages are filled with the results after the `query()` has been ran.

Even though you've said `->paginate(5)`, if some of those results were not found as a result of your `query()` then the page will not be filled up with 5 results.

If `search()` returned these 5 results for the page:
```
$users = [
	[
		'name' => 'John Smith',
		'location_id' => 1
	],[
		'name' => 'John Apple',
		'location_id' => 2
	],[
		'name' => 'John Wright',
		'location_id' => 2
	],[
		'name' => 'John Wrong',
		'location_id' => 1
	],[
		'name' => 'John Wick',
		'location_id' => 2
	]
]
```

The paginator thinks it has 5 results on the page, but in reality `$users` only contains:

```
$page = [
	[
		'name' => 'John Apple',
		'location_id' => 2
	],[
		'name' => 'John Wright',
		'location_id' => 2
	],[
		'name' => 'John Wick',
		'location_id' => 2
	]
]
```

**Comment by driesvints** [comment]
Does it also break if you use the Algolia driver?

**Comment by bkilshaw** [comment]
Yes, just tested it with the Algolia driver and I'm running into the same issue.

**Comment by driesvints** [comment]
@bkilshaw ok thanks for confirming that. Can you create a repo for me that reproduces the issue? Please commit any custom code separately from the base skeleton.

**Comment by driesvints** [comment]
Closing this issue because it's inactive, already solved, old or not relevant anymore. Feel free to reply if you're still experiencing this issue and we'll re-open this issue.

**Comment by jonnywilliamson** [comment]
Hi Dries! I took on that challenge! I was having the same (almost) issue, so I created a repo for you to see what we're talking about.

### Setup
**Repo to clone:**
https://github.com/jonnywilliamson/scoutbug

This is a brand new Laravel 8 app, with only `laravel scout` and `algolia/algoliasearch-client-php` added. Nothing else.

**`.env` file**
Only thing added was database name AND algolia credentials. If you're using another driver for your scout setup I'm sure you of all people will know how to set that up! 😆 
```
SCOUT_DRIVER=algolia
ALGOLIA_APP_ID=4xxxxxxxxxxxN
ALGOLIA_SECRET=fxxxxxxxxxxxxxxxxxxxxx4
```

Clone into folder and run migration. It will create only one extra table, `Articles`. Articles is a simple table that only has 2 main fields, `description` and `group_id`.

Run the `artisan DB:seed` command to add 8 records into the table. You can see the seeder data in `DatabaseSeeder.php`

**Index the records**
`artisan scout:import "\App\Models\Article"`

Setup is complete.

**The bug**

The 8 articles are very simple. 

- 5 articles have `group_id` of 1. 
- 3 have `group_id` of 2.

The word `YELLOW` appears only twice in all articles. One with a `group_id` of 1, one with a `group_id` of 2.


If you look in the `web.php` routes file, you'll see two route closures. 

Hitting the first `/search` route does a search for the word `yellow`. It theh paginates the result and returns it to the browser. As expected you will see the 2 articles  with the word yellow. The total field in the paginator will show 2.

![image](https://user-images.githubusercontent.com/2513663/114097459-f57d5180-98b7-11eb-855c-70717dad8d7c.png)



If you look at the `Article.php` model, you'll see a very simple scope in there (`groupTwo()`). It only limits the `group_id` to be equal to 2.

Now if we look at the second route in the `web.php` file `/scopesearch` you can see that we search for the word yellow again, only this time we then add a query scope (groupTwo()) before paginating the result.

This time the results are correct. We only have have ONE article returned....the one with `group_id` = 2 however the paginator shows the the TOTAL results returned is 2 which is incorrect.

![image](https://user-images.githubusercontent.com/2513663/114097775-6de41280-98b8-11eb-9669-b09f2578611a.png)


**Ideas**

I think the issue is here:
https://github.com/laravel/scout/blob/8.x/src/Builder.php#L274

The total is calculated from the rawresults got back from the search before any query is applied to them, rather than AFTER any queries.

Do I know how to solve that? Nope! 🤣 

Anyway. I hope that might help in reproducing the bug.

**Comment by jonnywilliamson** [comment]
@driesvints Apologies. I honestly don't know if I reply to a closed thread does it even appear in your notifications. I didn't want to type that all out and then you never see it. If this @ sends you a second one I'm very sorry!

**Comment by driesvints** [comment]
Investigating this.

**Comment by driesvints** [comment]
So, this cannot ever work because the additional complexity of scopes cannot be translated into a way a search index understands. You see, "where" clauses that are added directly to the search builder are translated to the engine underneath:

```php
Article::search('Yellow')
        ->where('group_id', 2)
        ->paginate(4);
```

This works because the engine supports it. But the following doesn't because the "builder" instance on the callback is a database query builder and not a search engine query builder. Their functionality differs and a search engine isn't a relational database:

```php
Article::search('Yellow')
        ->query(function(Builder $query){
            return $query->where('group_id', 2);
        })
        ->paginate(4);
```

In fact, there's a note about this in the docs:

![Screenshot 2021-04-09 at 15 28 37](https://user-images.githubusercontent.com/594614/114188203-32ddef80-9949-11eb-999a-499741a22eff.jpg)

After diving deep into this I wonder why the `query` method on the engine builders even exist. I don't see how it contributes anything. It's not even documented and it can only lead to mis-usage as above. It was added here by @taylorotwell: https://github.com/laravel/scout/commit/9f06b9c4d77879d53cf1680740b7f8bf19f78279

The way Scout works is that first, the engine performs the search query and then retrieves models based on the ID's it finds from the search query. Since the pagination has already been done through the search engine it cannot be modified afterwards with the database query upon model retrieval or you'll run into the above problems. 

I hope this explains it somehow. I hope you understand that there isn't any way for us to ever make this works since a database query and search engine just vastly differ from each other.

**Comment by jonnywilliamson** [comment]
Hi Dries, 

Thanks for taking the time to investigate and provide a detailed response and explanation.

![](https://i.kym-cdn.com/photos/images/original/000/918/126/241.gif)

Just kidding! I'll have to rethink my strategy on search!

**Comment by jonnywilliamson** [comment]
In the extreme chance anyone is following along and wants to know a possible way around this, the following is what is working for me now and I'm happy with it.

Basically I use a NORMAL eloquent builder so I have access to all the power of scopes etc that I want to use. Then I use a whereIn and only match it to the ID's returned from the search (using the keys method).

It's probably a little more expensive. But this is fine for my needs at the moment

```php
Article::groupTwo()
->whereIn('id', Article::search('YELLOW')->keys())
->paginate(4)
```

**Comment by driesvints** [comment]
@jonnywilliamson that's definitely an interesting solution and I think that'll work out nicely for your use case. But indeed it's a little bit more expensive as you're loading in all the keys the engine can find first and then performing the pagination on the Eloquent query side. You'll lose the performance benefit of doing the pagination on the engine side.

**Comment by bkilshaw** [comment]
This is the same solution I came up with originally. However; I use Postgres as my database and did some experimenting and was able to build a significantly faster search directly in Postgres so I went that route. 

One thing to be careful of is the number of records you match in your search results. I’m not sure how many records you’re working with, or what database you’re using, but there is a max number of IDs you can pass into the whereIn()

On Sat, Apr 10, 2021, at 1:25 AM, Jonathan wrote:
> 

> In the extreme chance anyone is following along and wants to know a possible way around this, the following is what is working for me now and I'm happy with it.

> Basically I use a NORMAL eloquent builder so I have access to all the power of scopes etc that I want to use. Then I use a whereIn and only match it to the ID's returned from the search (using the keys method).

> It's probably a little more expensive. But this is fine for my needs at the moment

> Article::groupTwo()
->whereIn('id', Article::search('YELLOW')->keys())
->paginate(4)

**Comment by driesvints** [comment]
Good news all. This will be solved in today's 9.0 release: https://github.com/laravel/scout/pull/469

**Comment by jonnywilliamson** [comment]
Ah thanks very much for the heads up Dries and thank you @crynobone for the work you did on this!
